cmd/keymasterd/2fa_bootstrapOTP.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/Cloud-Foundations/keymaster/lib/instrumentedwriter"
+	"github.com/Cloud-Foundations/keymaster/lib/webapi/v0/proto"
+)
+
+const bootstrapOtpAuthPath = "/api/v0/bootstrapOtpAuth"
+
+func (state *RuntimeState) BootstrapOtpAuthHandler(w http.ResponseWriter,
+	r *http.Request) {
+	if state.sendFailureToClientIfLocked(w, r) {
+		return
+	}
+	switch r.Method {
+	case "GET":
+		state.logger.Debugf(3, "Got client GET connection")
+		if err := r.ParseForm(); err != nil {
+			state.logger.Println(err)
+			state.writeFailureResponse(w, r, http.StatusBadRequest,
+				"Error parsing form")
+			return
+		}
+	case "POST":
+		state.logger.Debugf(3, "Got client POST connection")
+		if err := r.ParseForm(); err != nil {
+			state.logger.Println(err)
+			state.writeFailureResponse(w, r, http.StatusBadRequest,
+				"Error parsing form")
+			return
+		}
+	default:
+		state.writeFailureResponse(w, r, http.StatusMethodNotAllowed, "")
+		return
+	}
+	authUser, currentAuthLevel, err := state.checkAuth(w, r, AuthTypeAny)
+	if err != nil {
+		state.logger.Debugf(1, "%v", err)
+		return
+	}
+	w.(*instrumentedwriter.LoggingWriter).SetUsername(authUser)
+	var inputOTP string
+	if val, ok := r.Form["OTP"]; ok {
+		if len(val) > 1 {
+			state.writeFailureResponse(w, r, http.StatusBadRequest,
+				"Just one OTP Value allowed")
+			state.logger.Printf("Login with multiple OTP Values")
+			return
+		}
+		inputOTP = val[0]
+	}
+	requiredOTP := state.userBootstrapOtp(authUser)
+	if requiredOTP == "" {
+		state.writeFailureResponse(w, r, http.StatusPreconditionFailed,
+			"No valid Bootstrap OTP saved")
+		return
+	}
+	if inputOTP != requiredOTP {
+		state.logger.Printf("Invalid Bootstrap OTP value for %s\n",
+			authUser)
+		state.logger.Debugf(0, "  input: \"%s\" required: \"%s\"\n",
```

Comment:
Since you are printing something from the DB, the debug level must be much higher.

Reply:
Done.